### DISCUSSION TITLE
Scope & Uninstalling

### DISCUSSION THREAD

**Discussion by davidlamma** [microsoft/winget-cli#4089] [Q&A]
Hello, 

I am a smooth brained person tasked to set up new devices at a local MSP. I am currently working on batch script that focuses on mostly uninstalling applications using winget. I had a few questions about how `Scope` worked.

- Can I use `Scope` to uninstall applications entirely from the machine? By that I mean when I run my winget script as an admin and create a new user afterwards, will the applications remain uninstalled? I feel like the answer is yes? If that is not the case, is there any way of getting the desired outcome with winget? if the answer is no, I guess you can skip the questions below.

- I added the `Scope` "setting" into the .json, changed it to `uninstall behavior`  instead of `install behavior`, and the file is in the correct location. Do I need to call out the `-- scope`  `machine `  when I run this example command `winget uninstall -- id XXXX.XXX`? If so, where does it go? I tried messing with it a few different ways and I feel like I got different results every time. Looking through previous research, `Scope` is added in different places.  

- I don't think I understood the `Purge` setting. Would this help me with the quest for uninstalling?

Again, super sorry, but I am learning as I am going and this has been a bit of a roadblock for me. Some guidance would really be nice. 

Thanks

**Comment by Trenly** [comment]
> Can I use Scope to uninstall applications entirely from the machine? By that I mean when I run my winget script as an admin and create a new user afterwards, will the applications remain uninstalled? I feel like the answer is yes? If that is not the case, is there any way of getting the desired outcome with winget? if the answer is no, I guess you can skip the questions below.

WinGet can only affect what is currently on the machine. If you create a new user, it would have all the applications that a default user profile would have. Similarly, if you install an application and then create a new user, the new user would only have access to that application if a default user would have access to the application. At it's core, all WinGet does is download and run the installer for a program, or call the uninstaller the program has registered.

> I added the Scope "setting" into the .json, changed it to uninstall behavior instead of install behavior, and the file is in the correct location. Do I need to call out the -- scope machine  when I run this example command winget uninstall -- id XXXX.XXX? If so, where does it go? I tried messing with it a few different ways and I feel like I got different results every time. Looking through previous research, Scope is added in different places.

According to the Settings schema, the only supported property of `UninstallBehavior` is `purgePortablePackage`. This means you would need to call out `--scope` every time you run an uninstall. It can go anywhere in the command, there should be no difference between `winget uninstall --scope machine --id xxx.xxx` and `winget uninstall --id xxx.xxx --scope machine`. As for the "different results every time", my best guess (without knowing your exact configuration) is that various applications handle scope differently. Some applications only support a user scope, others only support machine scope, and some support both. Then there's weird ones like `Git.Git` which will always install in machine scope if the user is a member of the administrators group, regardless of whether the install is being run from an elevated context or not. Again, all WinGet does is call the uninstaller for the program. If I recall correctly, the `--scope` parameter for uninstall only limits the check for install scope - so if a program (like Git) claims to be installed as a user scope, but is actually installed in machine scope, then passing `--scope user` would remove it from all users. It's really all defined by how the installer and uninstaller for each program work.

https://github.com/microsoft/winget-cli/blob/231bb787eb6822b06f2921cddf9a9dfb60ef3ec3/schemas/JSON/settings/settings.schema.0.2.json#L183-L193

> I don't think I understood the Purge setting. Would this help me with the quest for uninstalling?

This setting only applies to packages where the InstallerType is `portable`. These are packages like ffmpeg where there is no installer, it's just a binary that can be run. When winget "installs" portable packages, it places them into their own folder and then adds the folder, or a symlink to the executable, to PATH. When uninstalling, WinGet will delete any items in the folder which were placed there by WinGet and leave any other files untouched, since some applications create files (e.g user settings, outputs from the program, etc.) when they run. The `--purge` option and corresponding `purgePortablePackage` setting tells WinGet to delete the folder entirely, including items which were not created by WinGet.
